File: implementations/fastapi/examples/src/order/domain/repository.py
from abc import ABC, abstractmethod

from .order import Order


class OrderRepository(ABC):

    @abstractmethod
    async def find_by_id(self, order_id: str) -> Order | None: ...

    @abstractmethod
    async def find_all(
        self,
        page: int,
        take: int,
        user_id: str | None = None,
        status: list[str] | None = None,
    ) -> tuple[list[Order], int]: ...

    @abstractmethod
    async def save(self, order: Order) -> None: ...

    @abstractmethod
    async def delete(self, order_id: str) -> None: ...

File: implementations/go/examples/internal/domain/order/repository.go
package order

import "context"

type FindQuery struct {
	Page   int
	Take   int
	UserID string
	Status []string
}

type Repository interface {
	FindByID(ctx context.Context, orderID string) (*Order, error)
	FindAll(ctx context.Context, q FindQuery) ([]*Order, int, error)
	Save(ctx context.Context, order *Order) error
	Delete(ctx context.Context, orderID string) error
}
